README.md
# Climatic - Python 3.x.x Temperature and Humidity Analyzer


The goal of this project is to create a single Flask web page that parses CSV
files containg temperature and humidity data. Users can upload their own CSV 
file or use the CSV file created by the logger and hosted by the web server.
The web page will take the CSV file from the user, parse its contents, and 
(if valid) display information about the contents of the file. The file 
will be in CSV format with "\r\n" line endings (Excel format). The data in the 
CSV file will contain the current temperature, humidity, date and time of 
collection, and any other pertinent information that can be gathered and 
stored by either additional sensors or by the Raspberry Pi itself. 
An example of the data expected by the project is included in the 
example.csv file.

The goal is to perform the CSV file parsing and any other calculations 
entirely on the client side. This will simplify the project in a way because 
the server will not have to handle file uploads and unexpected input 
from untrusted sources.

Currently, the logging component is separate from the server to allow it to be 
run independently. The temperature and humidity logging program is in the 
collect_data folder. Instructions for setting up and running the data logger 
is included in the collect_data README.md. 

**The Software**

The software will use Python 3.x.x and Flask for the web server and WiringPi 
for the Python code that interacts with the DHT11 temperature & humidity 
sensor. The web server will use Bootstrap to style the page, jQuery to 
assist with the JavaScript, and possibly Flot or Morris.js to create 
pretty looking graphs, tables, line charts, etc. 

**To Setup or Run**

First, get a copy of this code

    git clone https://github.com/spcnvdr/climatic.git

Change into the web directory with cd

    cd climatic/src

Create the virtual environment if this is the first run

    python3 -m venv venv

Enable the virtual environment:

    source venv/bin/activate

Install any Python 3.x.x dependencies

    pip3 install -r requirements.txt 

Then run the main Python 3.x.x script

    ./main

Then open a web browser to port 5000 on the localhost:

    127.0.0.1:5000

By default, the web server will listen on any public IP address and can be 
reached by any device on the same network, so be careful!


**To Do**

- [x] Determine the exact format of CSV files to use
- [ ] Finish the web page that will parse CSV files and display information
- [ ] ~~Make it possible to change the username and password~~
- [ ] Remove old, unneeded templates
- [ ] Use toasts for all alerts
- [x] Add space between menu bar and content on all pages, not just home
- [x] Manage JavaScript dependencies in a conventional manner (using npm)
- [ ] Update README with information on installing required npm packages
- [ ] Put custom CSS in a separate CSS file instead of using style attribute
- [ ] Use Flot.js for the line graph instead of Morris.js


**Contributing**

Pull requests, new feature suggestions, and bug reports/issues are
welcome.


**Versioning**

This project uses semantic versioning 2.0. Version numbers follow the
MAJOR.MINOR.PATCH format.


**License**

This project is licensed under the 3-Clause BSD License also known as the
*"New BSD License"* or the *"Modified BSD License"*. A copy of the license
can be found in the LICENSE file. A copy can also be found at the
[Open Source Institute](https://opensource.org/licenses/BSD-3-Clause)


**Acknowledgments**

This project uses many other open source projects. This project would
not be possible without the hard work of other developers. Below is a 
list of projects used by Climatic and links to their respective 
websites are given as well. This list is NOT exhaustive and in no particular 
order. The list below is subject to change without notice. 

Project - Website

* Python 3.x.x - https://www.python.org/

* Flask - https://flask.palletsprojects.com/en/1.1.x/

* Bootstrap 4 - https://getbootstrap.com/

* Popper.js - https://popper.js.org/

* Popper.js - https://popper.js.org/

* jQuery - https://jquery.com/

* Morris.js - https://morrisjs.github.io/morris.js/

* Raphael - https://dmitrybaranovskiy.github.io/raphael/

* Glyphicons (As part of Bootstrap 4) - https://www.glyphicons.com/

* Font Awesome - https://fontawesome.com/


**Security**

All commits will be signed with the following GPG key:
pub   rsa4096 2019-03-30 [SC]
      6E14 F46A EAA6 AE4B EFD5  8AAD 6E42 17B5 682C C50A
uid           [ultimate] Bryan Hawkins <spcnvdrr@protonmail.com>
sub   rsa4096 2019-03-30 [E]

The fingerprint for the signing key is:
6E14F46AEAA6AE4BEFD58AAD6E4217B5682CC50A


changelog
climatic (0.4.2) urgency=high;
  * Allow up to 150 data points on the graph to show trends over a longer 
    period of time. Temporary change until switch to Flot.js

  * Fixed template error in Flask that caused base template to be extended 
    incorrectly
  * Removed unused route in main.py
  * Added option to collect_data.py that automatically starts a new CSV file 
    every 24 hours

  -- Bryan Hawkins <spcnvdrr@protonmail.com> Tue, 23 Feb 2021 12:11:38 -0600


climatic (0.4.1) urgency=high;
  * Fixed a stupid bug in climatic.js that could return bad statistics if 
    a CSV file is uploaded with two data points with the first data point 
    being smaller in temperature or humidity. This was caused by using an 
    if-else instead of two separate if statements when looking for the 
    highest and lowest temperature/humidity.
  * Updated README.md to correctly describe the format of CSV file expected
  
  -- Bryan Hawkins <spcnvdrr@protonmail.com> Thu, 31 Dec 2020 08:49:43 -0600


climatic (0.4.0) urgency=high;
  * Use NPM to manage front end dependencies
  * Updated TO DO list in README.md
  * Remove all old front end dependencies and expect them to be installed 
    with NPM
  * Correct function comment above makeGraph() in climatic.js
  * Fix template.html due to changes in Font Awesome icon names between the 
    version served via NPM and the version downloaded directly from the 
    project's webpage
  * Remove a bunch of old static files no longer needed, templates from the 
    old Flask project this project started from, and other files hanging around
    that will not be used

  -- Bryan Hawkins <spcnvdrr@protonmail.com> Thu, 31 Dec 2020 07:57:40 -0600


climatic (0.3.3) urgency=low;
  * Modified climatic.js to only parse CSV data once
  * Opt for jQuery instead of JavaScript in climatic.js
  * Added Font Awesome to provide icons
  * Modified the style of the nav bar in template.html and removed redundant
    links
  * Updated TO DO list in README.md

  -- Bryan Hawkins <spcnvdrr@protonmail.com> Wed, 30 Dec 2020 13:42:06 -0600
  

climatic (0.3.2) urgency=high;
  * Correctly stop scheduled jobs when Ctrl-c interrupt is sent to collect_data
  * Move the apscheduler variable into main to reduce the number of global
    variables

  -- Bryan Hawkins <spcnvdrr@protonmail.com> Wed, 30 Dec 2020 12:48:39 -0600


climatic (0.3.1) urgency=low;
  * Modify collect_data to allow a custom collection interval to be specified
    with the -t/--time command line option.
  * Updated collect_data README.md
  * Moved the code in collect_data the creates the CSV file if it doesn't exist
    to main so that it is only run once.
  * Move the verbose print statement from the collect() function to the 
    timed_job() function.
  * Ran collect_data through the Python style checker Flake8 and fixed any 
    issues

  -- Bryan Hawkins <spcnvdrr@protonmail.com> Wed, 30 Dec 2020 12:35:36 -0600


climatic (0.3.0) urgency=low;
  * Got the basic features of the CSV parsing web page working.
  * Validate uploaded CSV files are in the same format as example.csv
  * Parse the CSV data and display the raw data in a card at the bottom of the 
    page
  * Added a line graph card made with Morris.js, but only show the 15 most 
    recent entries if there is a lot of data points
  * Added a statistics card that displays maximum and minimum temperature 
    and humidity. Also show the date and time when the max/min temp/humidity
    was seen. Also calculate and display average temperature and humidity.
  * Hide all cards on error
  * Replaced generic text in the file upload card with instructions
  * Fixed missing text in the displayed statistics
 
  -- Bryan Hawkins <spcnvdrr@protonmail.com> Tue, 29 Dec 2020 15:33:43 -0600


climatic (0.2.5) urgency=low;
  * Fixed misspellings and various errors in comments and documentation
 
  -- Bryan Hawkins <spcnvdrr@protonmail.com> Tue, 22 Dec 2020 12:03:45 -0600


climatic (0.2.4) urgency=low;
  * When creating a CSV file, include column definitions as the first row
  * Update collect_data README.md
 
  -- Bryan Hawkins <spcnvdrr@protonmail.com> Tue, 22 Dec 2020 11:32:22 -0600


climatic (0.2.3) urgency=low;
  * Update README.md
 
  -- Bryan Hawkins <spcnvdrr@protonmail.com> Mon, 21 Dec 2020 13:17:44 -0600


climatic (0.2.2) urgency=low;
  * Removed old code
  * Fixed incorrect information in the collect_data README.md
 
  -- Bryan Hawkins <spcnvdrr@protonmail.com> Mon, 21 Dec 2020 13:09:11 -0600


climatic (0.2.1) urgency=high;
  * Fixed bug introduced by removing old code.
 
  -- Bryan Hawkins <spcnvdrr@protonmail.com> Mon, 21 Dec 2020 12:29:11 -0600


climatic (0.2.0) urgency=med;
  * Updated data collection program, now stores temperature in Fahrenheit too.
  * Added an example CSV file to demonstrate the data produced by 
    the collect_data.py logger
  * Update README.md
  * Remove unnecessary or old code
  * Fixed previous changelog entry that incorrectly used tabs instead of 
    spaces
  * Fixed spelling issues

  -- Bryan Hawkins <spcnvdrr@protonmail.com> Mon, 21 Dec 2020 12:09:08 -0600


climatic (0.1.0) urgency=med;
  * Added the data collector program which continually runs, collects
    temperature and humidity data, and appends it to a CSV file.

  -- Bryan Hawkins <spcnvdrr@protonmail.com> Sun, 13 Dec 2020 13:41:15 -0600


climatic (0.0.6) urgency=high;
  * Renamed code folder to src to resolve conflict with Werkzeug package

  -- Bryan Hawkins <spcnvdrr@protonmail.com> Wed, 02 Dec 2020 14:39:47 -0600


climatic (0.0.5) urgency=high;
  * Reverted the previous pull request because of the numerous conflicts made
    with the main branch. Specifically, all the progress made on the Flask code
    was undone by the pull request. This was reverted back to the way the
    repository way in tag 0.0.3.
  * Removed a bad line from the requirements.txt file

  -- Bryan Hawkins <spcnvdrr@protonmail.com> Mon, 30 Nov 2020 14:14:13 -0600



tests/conftest.py
import pytest
from src.main import app

@pytest.fixture(scope='module')
def www():
    flask_app = app
    testing_client = flask_app.test_client()
    ctx = flask_app.app_context()
    ctx.push()
    yield testing_client
    ctx.pop()


tests/test_CSVController.py
def test_upload_csv(www):
    file = "example.csv"
    data = {
        'csv_file': (open(file, 'rb'), file)
    }
    response = www.post('/upload/csv', data=data)
    assert response.status_code == 200


src/app.py
#!/usr/bin/env python3
###########################################################################
# Stupid basic, ugly website that allows logins to view a secret web page #
# Probably not secure                                                     #
###########################################################################
from flask import Flask, render_template, redirect, url_for, request, \
    session, flash
from functools import wraps
from datetime import datetime
import os

app = Flask(__name__)

app.storage_path = os.path.join(os.path.realpath(__file__), '..', 'storage')
app.secret_key = "ls[F4U3yYkzI#%^wfupqXC3@fXSp"

from src.controllers import CSVController

def login_required(f):
    @wraps(f)
    def wrap(*args, **kwargs):
        if 'logged_in' in session:
            return f(*args, **kwargs)
        else:
            if request.path not in ["/login", "/logout"]:
                flash('You need to login first.')
            return redirect(url_for('login'))
    return wrap


@app.route("/")
@login_required
def home():
    now = datetime.now()
    date_str = now.ctime()
    if session["logged_in"]:
        return render_template("home.html", date=date_str)
    else:
        return render_template("login.html")


@app.route("/login", methods=["GET", "POST"])
def login():
    error = None
    if request.method == "POST":
        if request.form["username"] != "admin" or \
                request.form["password"] != "admin":
            error = "Invalid credentials"
        else:
            session["logged_in"] = True
            flash("You have successfuly logged in")
            return redirect(url_for("home"))
    return render_template("login.html", error=error)


@app.route("/logout")
@login_required
def logout():
    session.pop("logged_in", None)
    flash("You have logged out")
    return redirect(url_for('login'))


@app.route("/sandbox")
@login_required
def sandbox():
    now = datetime.now()
    date_str = now.ctime()
    return render_template("sandbox.html", date=date_str)

# Test page route
@app.route("/testpage")
@login_required
def testpage():
    # Return the rendered HTML template
    return render_template("testpage.html")


@app.route("/theme/toggle", methods=["POST"])
def theme_toggle():
    mode = request.form.get('mode', 'light')
    session['theme'] = mode
    return mode



if __name__ == "__main__":
    app.run(host="0.0.0.0", debug=True)


src/main.py
#!/usr/bin/env python3
###########################################################################
# Start of Flask web page                                                 #
###########################################################################
from flask import Flask, render_template, redirect, url_for, request, \
    session, flash
from functools import wraps
from datetime import datetime


app = Flask(__name__)

app.secret_key = "ls[F4U3yYkzI#%^wfupqXC3@fXSp"


def login_required(f):
    @wraps(f)
    def wrap(*args, **kwargs):
        if 'logged_in' in session:
            return f(*args, **kwargs)
        else:
            if request.path not in ["/login", "/logout"]:
                flash('You need to login first.')
            return redirect(url_for('login'))
    return wrap


@app.route("/")
def home():
    now = datetime.now()
    date_str = now.ctime()
    return render_template("home.html", date=date_str)


if __name__ == "__main__":
    app.run(host="0.0.0.0", debug=True)


src/controllers/CSVController.py
from main import app
from flask import request, redirect, flash, render_template
from werkzeug.utils import secure_filename
import os

def has_allowed_ext(filename, extensions):
    # print(filename.rsplit('.', 1))
    return '.' in filename and filename.rsplit('.', 1)[1].lower() in extensions

@app.route('/test', methods=['GET'])
def test():
    return render_template("csv.html")

@app.route('/upload/csv', methods=['POST'])
def upload_csv():
    if 'csv_file' in request.files:
        file = request.files['csv_file']
        # print(file.filename.rsplit('.', 1))
        if file and has_allowed_ext(file.filename, ('csv')):
            filename = secure_filename(file.filename)
            storage_location = os.path.join('storage', 'csv_files')
            if not os.path.exists(storage_location):
                os.makedirs(storage_location)

            file.save(os.path.join(storage_location, filename))
            return redirect(request.referrer)
    flash('errror')
    return redirect(request.referrer)
            #file.save(os.path.join('../storage'))


src/collect_data/collect_data.py
#!/usr/bin/env python3
###########################################################################
# This program interfaces with a DHT11 sensor to collect temperature      #
# and humidity data before appending it to a CSV file. The output file    #
# is created if it does not already exist. By default, the data is        #
# collected and appended to the CSV file every hour by default. In the    #
# future, add ability to change the measurement interval.                 #
#                                                                         #
# The CSV output creates the following columns:                           #
# timestamp, temperature in C, Temperature in F, humidity as a percentage #
#                                                                         #
# When looking at the DHT11 module from the front, the physical           #
# connections are as follows:                                             #
# DHT11 GND (pin 1) to Raspberry Pi ground                                #
# DHT11 DATA (pin 2) to Raspberry Pi pin 7                                #
# DHT11 VCC to 5 volts                                                    #
# NOTE: The above connection scheme is the exact same one as provided in  #
# the Kuman kit code examples. Check the documentation for the DHT11      #
# example for more information.                                           #
#                                                                         #
# To setup, run this command: pip3 install apscheduler                    #
# To run normally putting the output data in current directory:           #
#    ./collect_data.py                                                    #
# To run in background: ./collect_data.py &                               #
###########################################################################

import argparse
import time
import csv
from os.path import isfile
from datetime import datetime
from apscheduler.schedulers.blocking import BlockingScheduler
import RPi.GPIO as GPIO

# Default file name to store data in
CSV_FILE = "climate_data.csv"

# Columns definitions for the CSV file
COLUMNS = ["Timestamp", "Celsius", "Fahrenheit", "Humidity"]

# Number of seconds in an hour, collect temp/humidity every hour
HOUR = 3600

# The data collection interval in seconds, may be changed on command line
# with the -t/--time option
INTERVAL = HOUR

# Verbose and auto mode options
VERBOSE = False
AUTO = False

# Other globals
outfile = CSV_FILE

# save original filename for appending to in auto mode,
# and the time of the when we started the current CSV output file
original_file = ""
start_time = datetime.now()


# Create a new, unique file name based on the old file name
#  @param filename string the old file name
#  @returns a string consisting of the old filename with the date appended
#    to it
#
def new_filename(filename):
    newfilename = ""
    date = datetime.now()
    # append year, month, day - hour, minute to filename to make it unique
    strdate = date.strftime("%Y%m%d-%H%M")
    if "." in filename:
        temp = filename.split(".")
        newfilename = temp[0] + strdate + "." + temp[1]
    else:
        newfilename = filename + strdate
    return newfilename


# Roll the output file over to a new file
#  @param filename string the original filename
#  @returns the new output file name
#
def rollover_file(filename):
    global start_time

    newfile = new_filename(filename)
    while isfile(newfile):
        # TODO: Come up with a less shitty solution
        # If this file already exists, append Xs until we have a unique name
        if "." in newfile:
            temp = newfile.split(".")
            newfile = temp[0] + "X" + "." + temp[1]
        else:
            newfile += "X"

    # create the new file
    create_csv(newfile)
    # Reset the start time counter
    start_time = datetime.now()
    return newfile


# Convert a temperature in Celsius to Fahrenheit
#  @param temperature the temperature in Celsius to convert
#  @returns temperature as a float
#
def ctof(temperature):
    return ((temperature / 5) * 9) + 32


# Interface with DHT11 and return a list of current temp and humidity
#  @returns on success a list is returned which contains three items,
#  first is the temperature in Celsius, next is the temperature in Fahrenheit,
#  and the last item is the humidity as a percentage.
#
def collect():
    THdata = []
    channel = 7
    data = []

    # Set up GPIO pin and send start condition
    GPIO.setmode(GPIO.BOARD)
    time.sleep(2)
    GPIO.setup(channel, GPIO.OUT)
    GPIO.output(channel, GPIO.LOW)
    time.sleep(0.02)
    GPIO.output(channel, GPIO.HIGH)
    GPIO.setup(channel, GPIO.IN)

    # wait for Pi to finish sending start and
    # wait for ACK from DHT11
    while GPIO.input(channel) == GPIO.LOW:
        continue
    while GPIO.input(channel) == GPIO.HIGH:
        continue

    # Begin data transfer, data sent in 40 bit chunks
    j = 0
    while j < 40:
        k = 0
        while GPIO.input(channel) == GPIO.LOW:
            continue
        while GPIO.input(channel) == GPIO.HIGH:
            k += 1
            if k > 100:
                break
        if k < 8:
            data.append(0)
        else:
            data.append(1)
        j += 1

    humidity_bit = data[0:8]
    humidity_point_bit = data[8:16]
    temperature_bit = data[16:24]
    temperature_point_bit = data[24:32]
    check_bit = data[32:40]
    humidity = 0
    humidity_point = 0
    temperature = 0
    temperature_point = 0
    check = 0

    # Convert the raw bits into whole and fractional parts, e.g. 23.8C
    # compute checksum and store in the check variable
    for i in range(8):
        humidity += humidity_bit[i] * 2 ** (7 - i)
        humidity_point += humidity_point_bit[i] * 2 ** (7 - i)
        temperature += temperature_bit[i] * 2 ** (7 - i)
        temperature_point += temperature_point_bit[i] * 2 ** (7 - i)
        check += check_bit[i] * 2 ** (7 - i)
    tmp = humidity + humidity_point + temperature + temperature_point
    if check == tmp:
        THdata.append(str(temperature) + "." + str(temperature_point))
        # Convert C to F and append to returned data
        temp = float(THdata[0])
        f = "{:.4}".format(ctof(temp))
        THdata.append(f)
        THdata.append(str(humidity) + "%")
        return THdata
    else:
        # bad checksum, try again...
        time.sleep(1)
        return collect()


# Collect temperature and humidity data every hour
def timed_job():
    global outfile
    days_running = (datetime.now() - start_time).days
    if AUTO and days_running:
        outfile = rollover_file(original_file)

    data = []
    # create timestamp as first column
    data.append(str(datetime.now()))
    # collect data
    dhtdata = collect()
    temp_hum = float(dhtdata[2].strip("%"))
    while(temp_hum > 100):
        dhtdata = collect()
        temp_hum = float(dhtdata[2].strip("%"))

    # append temp in C, F, and humidity as a percentage to CSV file
    data.append(dhtdata[0])
    data.append(dhtdata[1])
    data.append(dhtdata[2])
    append_data(outfile, data)
    if VERBOSE:
        print("%s: temperature: %sC %sF\thumidity: %s" %
              (data[0], data[1], data[2], data[3]))


# Create a new CSV file with the correct column definitions
# called if the CSV file to append to does not exist
#  @param filename the name of CSV file to create
#
def create_csv(filename):
    with open(filename, "a", newline="") as fd:
        writer = csv.writer(fd)
        writer.writerow(COLUMNS)


# Append a row of data to the CSV file, creating it if necessary
#  @param filename string the name of the CSV file to open
#  @param data a list of strings containing the data to append
#  @returns number of bytes successfully stored in the CSV file
#
def append_data(filename, data):
    n = 0

    with open(filename, "a", newline="") as fd:
        f = csv.writer(fd)
        n = f.writerow(data)
    return n


# Main code
if __name__ == "__main__":
    # Set up argument parser
    argp = argparse.ArgumentParser("collect_data.py",
                                   description="Use DHT11 to collect and "
                                   "store temperature and humidity data in a "
                                   "CSV file\nConnect the DHT11 data pin to "
                                   "Raspberry Pi pin 7")
    argp.add_argument("-a",
                      "--auto", help="Automatically store data in a new CSV "
                      "file every 24 hours", action="store_true")
    argp.add_argument("-o",
                      "--output", help="Store data in a different output file")
    argp.add_argument("-t",
                      "--time", help="Collect data every N seconds",
                      metavar="N", type=int)
    argp.add_argument("-v",
                      "--verbose", help="Enable verbose mode",
                      action="store_true")

    # Parse arguments
    args = argp.parse_args()

    # If specified, change output file
    if args.output is not None:
        outfile = args.output

    # Enable verbose mode if specified
    if args.verbose:
        VERBOSE = True

    # If custom collection interval specified, use it
    if args.time:
        if args.time < 60:
            print("Error: time option CANNOT be less than 60 seconds!")
            exit(1)
        INTERVAL = args.time

    if args.auto:
        # enable auto mode, calculate when to roll over based on record count,
        # and save original filename for later modification
        AUTO = True
        original_file = outfile

    # If output file does not exist, create it
    if not isfile(outfile):
        create_csv(outfile)

    # Start the hourly task
    try:
        sched = BlockingScheduler()
        sched.add_job(timed_job, 'interval', seconds=INTERVAL)
        sched.start()
    except KeyboardInterrupt:
        print("Received Ctrl+c, exiting...")
        sched.shutdown()
        exit(0)
